### PR TITLE
Upstreaming fixes from refcount

### PR DESCRIPTION
**Pull request by d**
These are things in my way while working on a bigger patch set ("reference count refactor"). See each commit message for detail. I had to stack this branch on top of #11657 because Travis otherwise would fail. But reviewers can skip the commit titled "Use default GCC/Clang toolchain from Travis CI."

### CODE REVIEW

File: src/backend/gporca/libgpopt/src/base/CColConstraintsHashMapper.cpp [hunk 1]
```diff
@@ -17,17 +17,16 @@ CColConstraintsHashMapper::PdrgPcnstrLookup(CColRef *colref)
 
 // mapping between columns and single column constraints in array of constraints
 static ColRefToConstraintArrayMap *
-PhmcolconstrSingleColConstr(CMemoryPool *mp, CConstraintArray *drgPcnstr)
+PhmcolconstrSingleColConstr(CMemoryPool *mp, const CConstraintArray *drgPcnstr)
```
Comment (hardikar) [inline src/backend/gporca/libgpopt/src/base/CColConstraintsHashMapper.cpp:20]: Wait, how does this work? If `drgPcnstr` is const, how can we do `(*drgPcnstr)[ul]->AddRef()`?

**Review by hardikar — APPROVED** [review]
No objection to most of the changes, especially if it helps your tool. I don't think it reduces readability significantly. I do have one question/concern (see above). I checked the PR for sanity more than correctness - I'm gonna let the compiler/testing suite take care of that, there should be no changes because of this.

Thanks for getting rid of the 2 uber headers, that should help compile times a lot during development!

File: src/backend/gporca/libgpopt/src/base/CColConstraintsHashMapper.cpp (hunk 1 above)
Comment (d) [reply src/backend/gporca/libgpopt/src/base/CColConstraintsHashMapper.cpp:20]: Precisely! We _don't_ take ownership of this parameter @hardikar

Comment (hardikar) [reply src/backend/gporca/libgpopt/src/base/CColConstraintsHashMapper.cpp:20]: I mean, how can we call the non-const `operator []` on a const ptr `drgPcnstr`?

Comment (d) [reply src/backend/gporca/libgpopt/src/base/CColConstraintsHashMapper.cpp:20]: We _can't_ call a hypothetical non-`const` method (including overloaded member operators). However, this function only calls `const` member functions, including `operator []`

Comment (zellerh) [reply src/backend/gporca/libgpopt/src/base/CColConstraintsHashMapper.cpp:20]: This is NOT a suggestion for this PR, but maybe sometimes we could try the following in CDynamicPtrArray:

```
T * operator[](ULONG pos);
const T * operator[](ULONG pos) const;
```

Might be interesting to see whether we could get away with this without a lot of subsequent changes. That would make it more similar to std::array (https://en.cppreference.com/w/cpp/container/array/operator_at).

**Review by zellerh — COMMENTED** [review]
LGTM.

**Pull request merged**